### ISSUE TITLE
CAS SAML2 IdP does not send SAML2 LogoutResponse back to requesting service

### ISSUE THREAD

**Issue by gangyang7** [apereo/cas#2440]
When logging out from the logged-in SAML2 service and the service sent the SAML2 LogoutRequest to CAS SAML2 IdP, CAS SAML2 IdP failed to send LogoutResponse back to the service per SAML2 SLO profile.

The CAS tested were 5.1.0-RC1 and RC2 deployed in external Tomcat 8 on Windows 2008 R2.

To reproduce the problem:

- Build CAS SAML2 IdP with 5.1.0-RC2 using CAS War overlay
- Deploy the war into an external Tomcat 8
- Define only one SAML2 service in CAS SAML2 IdP. The service uses HTTP POST bindings for both SSO and SLO.
- Login to the SAML2 service via CAS SAML2 IdP
- Logout from the SAML2 service, which will trigger a SAML2 LogoutRequest sent to CAS SAML2 IdP via HTTP POST.
- Observe that CAS "Logout Successful" page is displayed instead of redirecting back to the service via HTTP POST.

A Tomcat console log is attached with CAS debug turned on for org.aperos. The log seem to indicate that the request body is empty. But the earlier log did show the LogoutRequest message, which is delivered in the HTTP request body.

Search for "!!! Start of SLO" when viewing the log.

[TomcatConsoleLog-NoRedirect.txt](https://github.com/apereo/cas/files/828524/TomcatConsoleLog-NoRedirect.txt)

**Comment by mmoayyed** [comment]
Thank you very much for this detailed report.

Have you managed to step into the relevant components via a debugger, etc to diagnose the issue?

**Comment by gangyang7** [comment]
You are welcome. I did not step into the code since I've just cloned the
code base and have not attempted to build it first. But I did look at the
relevant code (in master branch though) to see what the code is doing.
Specifically, I looked at the
org.apereo.cas.support.saml.authentication.principal.SamlServiceFactory
class where it complained about empty request body. I saw several things
that I have questions for:

- The code seemed to require either "TARGET" parameter and request body,
but the log message wording sounded like it require both.

- The SamlServiceFactory.createService() seemed to be only looking for
Artifact parameters in the request body. I was expecting it to look for
"SAMLRequest" parameter instead. Does CAS IdP support POST binding for SLO?

Gang

**Comment by mmoayyed** [comment]
Great work. Let me start by saying that you're not looking in the right places. You want to look into the LogoutManagerImpl, and the components in the saml-idp module, specially `SamlIdPSingleLogoutServiceLogoutUrlBuilder`. The `SamlServiceFactory` is not relevant here. And yes, you should look into the master branch. 

There are guides available on the project site that describe how you can build and set up a dev environment. LMK if you have any questions.

**Comment by mmoayyed** [comment]
This https://apereo.github.io/cas/developer/Build-Process.html and this https://apereo.github.io/cas/developer/Contributor-Guidelines.html should help some.

**Comment by gangyang7** [comment]
Apparently I'm not familiar with CAS's code. I looked into SamlServiceFactory
because it logged something that looked like the first sign of something
might have gone wrong. There was not much debug logs after that and
SamlIdPSingleLogoutServiceLogoutUrlBuilder wasn't even present in the log.


Gang

**Comment by mmoayyed** [comment]
There may not be logging, sure, but those are the components that should be involved. If they are not, then that's where the issue lies really.

**Comment by mmoayyed** [comment]
I tried to run through this sequence under RC3 SNAPSHOT and could not duplicate. Please try again and report back. We can further investigate.

**Comment by gangyang7** [comment]
How do I refer to RC3 SNAPSHOT in CAS War Overlay?

**Comment by mmoayyed** [comment]
How did you deploy RC1 or RC2? Maven?

**Comment by gangyang7** [comment]
Eclipse, Maven, CAS War Overlay.

**Comment by mmoayyed** [comment]
Cool. In your pom, find the setting that references your CAS version. Switch it to match this version instead:
https://github.com/apereo/cas/blob/master/gradle.properties#L24

Then review this to download and build snapshots:
https://github.com/apereo/cas-overlay-template#build

**Comment by gangyang7** [comment]
Followed your instructions and redeployed cas.war. There was no error, but cas service did not come up. The character-formed (CAS) logo with the system info were noticeable missing.

**Comment by mmoayyed** [comment]
Look into your tomcat logs please; there should be error entries somewhere.

**Comment by mmoayyed** [comment]
Also please share your pom.xml file here (Paste it if you can please. Much easier to review)

**Comment by gangyang7** [comment]
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd ">
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.apereo.cas</groupId>
    <artifactId>cas-overlay</artifactId>
    <packaging>war</packaging>
    <version>1.0</version>

    <build>
        <plugins>
            <plugin>
                <groupId>com.rimerosolutions.maven.plugins</groupId>
                <artifactId>wrapper-maven-plugin</artifactId>
                <version>0.0.4</version>
                <configuration>
                    <verifyDownload>true</verifyDownload>
                    <checksumAlgorithm>MD5</checksumAlgorithm>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <version>${springboot.version}</version>
                <configuration>
                    <mainClass>org.springframework.boot.loader.WarLauncher</mainClass>
                    <addResources>true</addResources>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-war-plugin</artifactId>
                <version>2.6</version>
                <configuration>
                    <warName>cas</warName>
                    <failOnMissingWebXml>false</failOnMissingWebXml>
                    <recompressZippedFiles>false</recompressZippedFiles>
                    <archive>
                        <compress>false</compress>
                        <manifestFile>${project.build.directory}/war/work/org.apereo.cas/cas-server-webapp/META-INF/MANIFEST.MF
                        </manifestFile>
                    </archive>
                    <overlays>
                        <overlay>
                            <groupId>org.apereo.cas</groupId>
                            <artifactId>cas-server-webapp</artifactId>
                        </overlay>
                    </overlays>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.3</version>
            </plugin>
        </plugins>
        <finalName>cas</finalName>
    </build>

    <dependencies>
        <dependency>
            <groupId>org.apereo.cas</groupId>
            <artifactId>cas-server-webapp</artifactId>
            <version>${cas.version}</version>
            <type>war</type>
            <scope>runtime</scope>
        </dependency>
		<dependency>
			<groupId>org.apereo.cas</groupId>
			<artifactId>cas-server-support-saml</artifactId>
			<version>${cas.version}</version>
			
		</dependency>
		<dependency>
			<groupId>org.apereo.cas</groupId>
			<artifactId>cas-server-support-saml-idp</artifactId>
			<version>${cas.version}</version>
		</dependency>
		<dependency>
		    <groupId>org.apereo.cas</groupId>
		    <artifactId>cas-server-support-json-service-registry</artifactId>
		    <version>${cas.version}</version>
		    
		</dependency>
    </dependencies>

    <properties>
        <cas.version>5.1.0-RC3-SNAPSHOT</cas.version>
        <springboot.version>1.4.2.RELEASE</springboot.version>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <repositories>
        <repository>
            <id>sonatype-releases</id>
            <url>http://oss.sonatype.org/content/repositories/releases/</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
            <releases>
                <enabled>true</enabled>
            </releases>
        </repository>
        <repository>
            <id>sonatype-snapshots</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
            <releases>
                <enabled>false</enabled>
            </releases>
        </repository>
        <repository>
            <id>shibboleth-releases</id>
            <url>https://build.shibboleth.net/nexus/content/repositories/releases</url>
        </repository>
        <repository>
            <id>spring-milestones</id>
            <url>https://repo.spring.io/milestone</url>
        </repository>
    </repositories>

    <profiles>
        <profile>
            <activation>
                <activeByDefault>false</activeByDefault>
            </activation>
            <id>pgp</id>
            <build>
                <plugins>
                    <plugin>
                        <groupId>com.github.s4u.plugins</groupId>
                        <artifactId>pgpverify-maven-plugin</artifactId>
                        <version>1.1.0</version>
                        <executions>
                            <execution>
                                <goals>
                                    <goal>check</goal>
                                </goals>
                            </execution>
                        </executions>
                        <configuration>
                            <pgpKeyServer>hkp://pool.sks-keyservers.net</pgpKeyServer>
                            <pgpKeysCachePath>${settings.localRepository}/pgpkeys-cache</pgpKeysCachePath>
                            <scope>test</scope>
                            <verifyPomFiles>true</verifyPomFiles>
                            <failNoSignature>false</failNoSignature>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        </profile>
    </profiles>
</project>

**Comment by gangyang7** [comment]
Removed all Tomcat logs and restarted Tomcat. Checked all logs w/o error. It's almost like deploying an empty cas.war. cas.war does not have web.xml.

**Comment by mmoayyed** [comment]
Ah, so first thing: 

This:

```
<springboot.version>1.4.2.RELEASE</springboot.version>
```

Must be:

```
<springboot.version>1.5.2.RELEASE</springboot.version>
```

Try that.

**Comment by mmoayyed** [comment]
If that fails, I'll take a pass later. TBH I have not run the overlays with 5.1 yet so chances are some additional changes are in order though shouldn't be too drastic.

**Comment by gangyang7** [comment]
Same - cas service did not come up.

**Comment by gangyang7** [comment]
Compared with RC2 cas.war, RC3-SNAPSHOT cas.war is missing org/springfamework/boot/loader/*. I assume it's responsible for loading and intialization, though I don't have much knowledge of Spring's Boot and how it works in an external Tomcat deployment.

**Comment by mmoayyed** [comment]
No that bit is only used when you run the embedded tomcat. Shouldnt apply to your case. 

Thank you for trying things out thus far. I'll take a pass and see what might be missing a bit later. Stay tuned.

**Comment by mmoayyed** [comment]
You should be able to test the fix in about an hour or so, by using the new SNAPSHOT.

**Comment by gangyang7** [comment]
The cas service is back to active. But it's trying to get configuration from http://localhost:8888/cas/native, which I don't have. ?? Here's the relevant Tomcat console logs:

2017-03-10 14:21:22,688 INFO [org.springframework.cloud.config.client.ConfigServicePropertySourceLoc
ator] - <Fetching config from server at: http://localhost:8888>
2017-03-10 14:21:23,826 WARN [org.springframework.boot.SpringApplication] - <Error handling failed (
ApplicationEventMulticaster 10-Mar-2017 14:21:23.838 SEVERE [localhost-startStop-1] org.apache.catal
ina.core.ContainerBase.addChildInternal ContainerBase.addChild: start:
 org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].Standar
dHost[localhost].StandardContext[/cas]]
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:154)
        at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:725)
        at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:701)
        at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:717)
        at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1101)
        at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1786)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: Could not locate PropertySource and the fail fast proper
ty is set, failing
        at org.springframework.cloud.config.client.ConfigServicePropertySourceLocator.locate(ConfigS
ervicePropertySourceLocator.java:130)
        at org.springframework.cloud.bootstrap.config.PropertySourceBootstrapConfiguration.initializ
e(PropertySourceBootstrapConfiguration.java:89)
        at org.springframework.boot.SpringApplication.applyInitializers(SpringApplication.java:611)
        at org.springframework.boot.SpringApplication.prepareContext(SpringApplication.java:348)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
        at org.springframework.boot.web.support.SpringBootServletInitializer.run(SpringBootServletIn
itializer.java:151)
        at org.springframework.boot.web.support.SpringBootServletInitializer.createRootApplicationCo
ntext(SpringBootServletInitializer.java:131)
        at org.springframework.boot.web.support.SpringBootServletInitializer.onStartup(SpringBootSer
vletInitializer.java:86)
        at org.springframework.web.SpringServletContainerInitializer.onStartup(SpringServletContaine
rInitializer.java:169)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5156)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        ... 10 more
Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "htt
p://localhost:8888/cas/native": Connection refused: connect; nested exception is java.net.ConnectExc
eption: Connection refused: connect
        at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:666)
        at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:613)
        at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:531)
        at org.springframework.cloud.config.client.ConfigServicePropertySourceLocator.getRemoteEnvir
onment(ConfigServicePropertySourceLocator.java:170)
        at org.springframework.cloud.config.client.ConfigServicePropertySourceLocator.locate(ConfigS
ervicePropertySourceLocator.java:91)
        ... 20 more
Caused by: java.net.ConnectException: Connection refused: connect
        at java.net.DualStackPlainSocketImpl.connect0(Native Method)
        at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source)
        at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
        at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
        at java.net.AbstractPlainSocketImpl.connect(Unknown Source)
        at java.net.PlainSocketImpl.connect(Unknown Source)
        at java.net.SocksSocketImpl.connect(Unknown Source)
        at java.net.Socket.connect(Unknown Source)
        at java.net.Socket.connect(Unknown Source)
        at sun.net.NetworkClient.doConnect(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.http.HttpClient.<init>(Unknown Source)
        at sun.net.www.http.HttpClient.New(Unknown Source)
        at sun.net.www.http.HttpClient.New(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.connect(Unknown Source)
        at org.springframework.http.client.SimpleBufferingClientHttpRequest.executeInternal(SimpleBu
fferingClientHttpRequest.java:78)
        at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(Abstra
ctBufferingClientHttpRequest.java:48)
        at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpReque
st.java:53)
        at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:652)
        ... 24 more

**Comment by mmoayyed** [comment]
https://apereo.github.io/cas/development/installation/Configuration-Server-Management.html

**Comment by gangyang7** [comment]
In earlier version, cas.serviceRegistry.config.location defines the IdP metadata location. So I have in my cas.properties:

cas.authn.samlIdp.metadata.location=file:///C:/temp/cas/idp

This does not seem to work any more. ?? The logs mentioned C:\etc\cas\* in several places, but I could not find where they are specified. I'm on Windows. The following is a relevant Tomcat console logs:

2017-03-10 15:32:14,788 WARN [org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebA
pplicationContext] - <Exception encountered during context initialization - cancelling refresh attem
pt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'shibbole
thIdpMetadataAndCertificatesGenerationService': Invocation of init method failed; nested exception i
s java.lang.IllegalArgumentException: Metadata directory location file [C:\etc\cas\saml] cannot be l
ocated/created>
2017-03-10 15:32:14,793 WARN [org.apereo.cas.services.ServiceRegistryConfigWatcher] - <Directory key
 is no longer valid. QuittinException in thread "org.apereo.cas.services.JsonServiceRegistryDao" g w
atcher service>
java.nio.file.ClosedWatchServiceException
        at sun.nio.fs.AbstractWatchService.checkOpen(Unknown Source)
        at sun.nio.fs.AbstractWatchService.checkKey(Unknown Source)
        at sun.nio.fs.AbstractWatchService.take(Unknown Source)
        at org.apereo.cas.services.ServiceRegistryConfigWatcher.run(ServiceRegistryConfigWatcher.jav
a:72)
        at java.lang.Thread.run(Unknown Source)
10-Mar-2017 15:32:14.881 SEVERE [localhost-startStop-1] org.apache.catalina.core.ContainerBase.addCh
ildInternal ContainerBase.addChild: start:

**Comment by mmoayyed** [comment]
> In earlier version, cas.serviceRegistry.config.location defines the IdP metadata location. 

How?!!

**Comment by mmoayyed** [comment]
Your logs show that CAS is trying to create  [C:\etc\cas\saml]. Which means it's not reading your config.

**Comment by gangyang7** [comment]
This only two config files I modified are bootstrap.properties and cas.properties. See attached. Do you see anything wrong? Thanks.

[cas.properties.txt](https://github.com/apereo/cas/files/835116/cas.properties.txt)
[bootstrap.properties.txt](https://github.com/apereo/cas/files/835117/bootstrap.properties.txt)

**Comment by gangyang7** [comment]
Probably easier to read:

bootstrap.properties:
-----------------------
##
# CAS Spring Cloud Config Client
#
# Name of the application for which environment settings and properties should be fetched.
# This should map to a cas.yml or cas.properties file.
spring.application.name=cas
spring.profiles.active=standalone

# Define where the configuration server is running
# and what credentials are used to access it.
# By default, basic authN is used.
spring.cloud.config.uri=http://casuser:Mellon@localhost:8888/casconfigserver
spring.cloud.config.profile=native
spring.cloud.config.server.native.searchLocations=file:///C:/Java/apache-tomcat-8.0.22/webapps/cas/WEB-INF/classes

# Configuration server is off by default.
spring.cloud.config.enabled=false
        
spring.cloud.config.watch.enabled=true
spring.cloud.config.watch.initialDelay=30000
spring.cloud.config.watch.delay=1000

spring.cloud.config.fail-fast=true
health.config.enabled=true


cas.properties:
----------------
cas.server.name: https://openam.example.org:8443
cas.server.prefix: https://openam.example.org:8443/cas

cas.adminPagesSecurity.ip=127\.0\.0\.1

#logging.config: file:/etc/cas/config/log4j2.xml
logging.config=classpath:log4j2.xml
# cas.serviceRegistry.config.location: classpath:/services
cas.serviceRegistry.config.location=file:///C:/temp/cas/services

# SAML 2 IdP:
cas.authn.samlIdp.metadata.location=file:///C:/temp/cas/idp
cas.authn.samlIdp.response.skewAllowance=10

**Comment by gangyang7** [comment]
bootstrap.properties:
-----------------------
\##
\# CAS Spring Cloud Config Client
\#
\# Name of the application for which environment settings and properties should be fetched.
\# This should map to a cas.yml or cas.properties file.
spring.application.name=cas
spring.profiles.active=standalone

\# Define where the configuration server is running
\# and what credentials are used to access it.
\# By default, basic authN is used.
spring.cloud.config.uri=http://casuser:Mellon@localhost:8888/casconfigserver
spring.cloud.config.profile=native
spring.cloud.config.server.native.searchLocations=file:///C:/Java/apache-tomcat-8.0.22/webapps/cas/WEB-INF/classes

\# Configuration server is off by default.
spring.cloud.config.enabled=false
        
spring.cloud.config.watch.enabled=true
spring.cloud.config.watch.initialDelay=30000
spring.cloud.config.watch.delay=1000

spring.cloud.config.fail-fast=true
health.config.enabled=true

cas.properties:
----------------
cas.server.name: https://openam.example.org:8443
cas.server.prefix: https://openam.example.org:8443/cas

cas.adminPagesSecurity.ip=127\.0\.0\.1

\#logging.config: file:/etc/cas/config/log4j2.xml
logging.config=classpath:log4j2.xml
\# cas.serviceRegistry.config.location: classpath:/services
cas.serviceRegistry.config.location=file:///C:/temp/cas/services

\# SAML 2 IdP:
cas.authn.samlIdp.metadata.location=file:///C:/temp/cas/idp
cas.authn.samlIdp.response.skewAllowance=10

**Comment by mmoayyed** [comment]
I'll take a look sure, but in the future please use ``` to format code. (I did one for the above, just as an example).

**Comment by mmoayyed** [comment]
You should probably remove your bootstrap file since there isnt anything in there different from what CAS provides (or I don't see anything different) 

Please attach DEBUG logs from the very start. Also step into the code, specially here https://github.com/apereo/cas/blob/master/core/cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/config/CasCoreBootstrapStandaloneConfiguration.java and you'll see how settings are loaded.

**Comment by gangyang7** [comment]
I only added this line to specify where to search for configs:

spring.cloud.config.server.native.searchLocations=file:///C:/Java/apache-tomcat-8.0.22/webapps/cas/WEB-INF/classes

Without it for earlier version, CAS looks into /etc/cas/. Things may have changed.

**Comment by mmoayyed** [comment]
Yes and that's failing. I encourage you to review this doc:
https://apereo.github.io/cas/development/installation/Configuration-Server-Management.html

**Comment by gangyang7** [comment]
This used to do the trick and as usual I've learned it from the searching internet w/o fully understanding what went behind the curtain. Apparently things changed. If you know what I'm missing and it's not too much to ask, please help. 

I know CAS may be aiming at a highly sophisticated configuration framework using Spring. But it's also supposed to be highly "automated" and easy to do the initial setup. Unfortunately it's been a very time-consuming task to configure CAS out of box for someone like me who's working on the client side and just wanted to be able support CAS server as another SAML 2.0 IdP.

**Comment by mmoayyed** [comment]
I strongly advise against making random changes based on recipes found on the web without understanding what they do and what they are for. That simply leads to disaster. Everything you need to know is documented in that link I provided above. Take a pass, read and digest...Or find someone who can help you better in a more detailed manner, or find commercial support to get timely and accurate support or find a better product with an easier setup and benefits.

**Comment by mmoayyed** [comment]
You may also find some value in this post https://apereo.github.io/2017/03/31/rc3-release/

**Comment by gangyang7** [comment]
This problem is not fixed in 5.1.0-RC3-SNAPSHOT as of today. Good thing is that issue 2439 is fixed and the missing SessionIndex AuthnStatement is fixed (I did not report it). But after logout, CAS displays the "logout successful" page w/o sending back LogoutResponse.

[tomcat-console-log.txt](https://github.com/apereo/cas/files/839651/tomcat-console-log.txt)
